### PR TITLE
ORC-759: StructBatchReader should always skip processing on the rootReader

### PR DESCRIPTION
**Pull request by pavibhai**
### What changes were proposed in this pull request?
StructBatchReader assumes rootReader has no nulls and skips rootReader processing in the following methods:
* seek
* skip
* startStripe

This behavior will be consistent with the behavior of nextBatch which already makes this assumption.

### Why are the changes needed?
Make the behavior consistent on all the methods.

### How was this patch tested?
Regression test, none of the existing tests have failed.

### CODE REVIEW

File: java/core/src/java/org/apache/orc/impl/reader/tree/StructBatchReader.java [hunk 1]
```diff
@@ -20,11 +20,18 @@
...
 import org.apache.orc.impl.TreeReaderFactory;
+import org.apache.orc.impl.reader.StripePlanner;
 
 import java.io.IOException;
 import java.util.Set;
 
+/**
+ * Handles the Struct rootType for batch handling. The handling assumes that the root
+ * {@link org.apache.orc.impl.TreeReaderFactory.StructTreeReader} has no nulls, this is required as
+ * the {@link VectorizedRowBatch} does not represent the root Struct as a vector.
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/reader/tree/StructBatchReader.java:33]: Nice, I would rephrase to something like: ..has no nulls. Root Struct vector is not represented as part of the final  the {@link  org.apache.hadoop.hive.ql.exec.vector.VectorizedRowBatch}.

File: java/core/src/java/org/apache/orc/impl/InStream.java [hunk 2]
```diff
@@ -621,9 +631,20 @@ private ByteBuffer slice(int chunkLength) throws IOException {
...
+     * Seek to the desired chunk based on the input position.
+     *
+     * @param desired position in the compressed stream
+     * @return Indicates whether a seek was performed or not
+     * @throws IOException when seeking outside the stream bounds
+     */
+    boolean seek(long desired) throws IOException {
       if (desired == 0 && bytes == null) {
-        return;
+        return false;
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/InStream.java:643]: Lets remove ORC-758 leftovers from the PR

File: java/core/src/java/org/apache/orc/impl/reader/tree/StructBatchReader.java [hunk 3]
```diff
@@ -76,4 +83,31 @@ public void nextBatch(VectorizedRowBatch batch, int batchSize) throws IOExceptio
       }
     }
   }
+
+  @Override
+  public void startStripe(StripePlanner planner) throws IOException {
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/reader/tree/StructBatchReader.java:88]: Makes sense, maybe we should make these methods abstract on BatchReader instead?
Then we would only keep setVectorColumnCount and resetBatch methods with implementation that apply both for Primitive and Struct reading.

File: java/core/src/test/org/apache/orc/impl/TestInStream.java [hunk 4]
```diff
@@ -434,8 +437,30 @@ public void testCompressed() throws Exception {
       int x = in.read();
       assertEquals(i & 0xff, x);
     }
+
+    // Forward seek test
```
Comment (pgaref) [inline java/core/src/test/org/apache/orc/impl/TestInStream.java:441]: another ORC-758 leftover

**Review by pgaref — COMMENTED** [review]
Almost there, left some comments @pavibhai

File: java/core/src/java/org/apache/orc/impl/reader/tree/StructBatchReader.java (hunk 1 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/reader/tree/StructBatchReader.java:33]: Changed

File: java/core/src/java/org/apache/orc/impl/reader/tree/StructBatchReader.java (hunk 3 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/reader/tree/StructBatchReader.java:88]: Good idea. Pushed the implementation into PrimitiveBatchReader.

Now only the methods `setVectorColumnCount` and `resetBatch` methods are implemented in BatchReader.

File: java/core/src/java/org/apache/orc/impl/InStream.java (hunk 2 above)
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/InStream.java]: Removed ORC-758 references

**Review by pgaref — APPROVED** [review]
Hey @pavibhai I am +1 on latest changes -- can merge after addressing the minor comment above :)

File: java/core/src/java/org/apache/orc/impl/reader/tree/PrimitiveBatchReader.java [hunk 5]
```diff
@@ -15,10 +15,14 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
```
[the remark this answers is not in the document]
Comment (pavibhai) [reply java/core/src/java/org/apache/orc/impl/reader/tree/PrimitiveBatchReader.java:18]: Sure removed

**Comment by pgaref** [comment]
Thanks @pavibhai ! This is now merged!

**Comment by dongjoon-hyun** [comment]
Thank you, @pavibhai and @pgaref .
